File: sdkwork-rtc-sdk-flutter/lib/src/rtc_provider_selection.dart
import 'rtc_provider_catalog.dart';

enum RtcProviderSelectionSource {
  provider_url,
  provider_key,
  tenant_override,
  deployment_profile,
  default_provider,
}

final class ParsedRtcProviderUrl {
  const ParsedRtcProviderUrl({
    required this.providerKey,
    required this.rawUrl,
  });

  final String providerKey;
  final String rawUrl;
}

final class RtcProviderSelection {
  const RtcProviderSelection({
    required this.providerKey,
    required this.source,
  });

  final String providerKey;
  final RtcProviderSelectionSource source;
}

final class RtcProviderSelectionRequest {
  const RtcProviderSelectionRequest({
    this.providerUrl,
    this.providerKey,
    this.tenantOverrideProviderKey,
    this.deploymentProfileProviderKey,
  });

  final String? providerUrl;
  final String? providerKey;
  final String? tenantOverrideProviderKey;
  final String? deploymentProfileProviderKey;
}

const List<String> rtcProviderSelectionSources = <String>[
  'provider_url',
  'provider_key',
  'tenant_override',
  'deployment_profile',
  'default_provider',
];

const List<String> rtcProviderSelectionPrecedence = <String>[
  'provider_url',
  'provider_key',
  'tenant_override',
  'deployment_profile',
  'default_provider',
];

bool _hasRtcProviderSelectionText(String? value) =>
    value != null && value.trim().isNotEmpty;

ParsedRtcProviderUrl parseRtcProviderUrl(String providerUrl) {
  final trimmed = providerUrl.trim();
  if (!trimmed.startsWith('rtc:') || !trimmed.contains('://')) {
    throw ArgumentError.value(providerUrl, 'providerUrl', 'Invalid RTC provider URL');
  }

  return ParsedRtcProviderUrl(
    providerKey: trimmed.substring(4).split('://').first.toLowerCase(),
    rawUrl: providerUrl,
  );
}

RtcProviderSelection resolveRtcProviderSelection(
  RtcProviderSelectionRequest request, {
  String defaultProviderKey = RtcProviderCatalog.DEFAULT_RTC_PROVIDER_KEY,
}) {
  if (_hasRtcProviderSelectionText(request.providerUrl)) {
    return RtcProviderSelection(
      providerKey: parseRtcProviderUrl(request.providerUrl!).providerKey,
      source: RtcProviderSelectionSource.provider_url,
    );
  }

  if (_hasRtcProviderSelectionText(request.providerKey)) {
    return RtcProviderSelection(
      providerKey: request.providerKey!.trim(),
      source: RtcProviderSelectionSource.provider_key,
    );
  }

  if (_hasRtcProviderSelectionText(request.tenantOverrideProviderKey)) {
    return RtcProviderSelection(
      providerKey: request.tenantOverrideProviderKey!.trim(),
      source: RtcProviderSelectionSource.tenant_override,
    );
  }

  if (_hasRtcProviderSelectionText(request.deploymentProfileProviderKey)) {
    return RtcProviderSelection(
      providerKey: request.deploymentProfileProviderKey!.trim(),
      source: RtcProviderSelectionSource.deployment_profile,
    );
  }

  return RtcProviderSelection(
    providerKey: defaultProviderKey,
    source: RtcProviderSelectionSource.default_provider,
  );
}

File: sdkwork-rtc-sdk-csharp/src/SDKWork.Rtc.Sdk/RtcProviderSelection.cs
namespace Sdkwork.Rtc.Sdk;

using System;
using System.Collections.Generic;

public enum RtcProviderSelectionSource
{
    provider_url,
    provider_key,
    tenant_override,
    deployment_profile,
    default_provider,
}

public sealed record ParsedRtcProviderUrl(string providerKey, string rawUrl);

public sealed record RtcProviderSelectionRequest(
    string? providerUrl = null,
    string? providerKey = null,
    string? tenantOverrideProviderKey = null,
    string? deploymentProfileProviderKey = null
);

public sealed record RtcProviderSelection(
    string providerKey,
    RtcProviderSelectionSource source
)
{
    public static readonly IReadOnlyList<string> RtcProviderSelectionSources =
    [
        "provider_url",
        "provider_key",
        "tenant_override",
        "deployment_profile",
        "default_provider",
    ];

    public static readonly IReadOnlyList<string> RtcProviderSelectionPrecedence =
    [
        "provider_url",
        "provider_key",
        "tenant_override",
        "deployment_profile",
        "default_provider",
    ];

    public static ParsedRtcProviderUrl ParseRtcProviderUrl(string providerUrl)
    {
        var trimmed = providerUrl.Trim();
        if (!trimmed.StartsWith("rtc:", StringComparison.OrdinalIgnoreCase) || !trimmed.Contains("://", StringComparison.Ordinal))
        {
            throw new ArgumentException($"Invalid RTC provider URL: {providerUrl}", nameof(providerUrl));
        }

        return new ParsedRtcProviderUrl(
            trimmed[4..trimmed.IndexOf("://", StringComparison.Ordinal)].ToLowerInvariant(),
            providerUrl
        );
    }

    public static RtcProviderSelection ResolveRtcProviderSelection(
        RtcProviderSelectionRequest? request = null,
        string defaultProviderKey = RtcProviderCatalog.DEFAULT_RTC_PROVIDER_KEY
    )
    {
        request ??= new RtcProviderSelectionRequest();

        if (HasText(request.providerUrl))
        {
            return new RtcProviderSelection(
                ParseRtcProviderUrl(request.providerUrl!).providerKey,
                RtcProviderSelectionSource.provider_url
            );
        }

        if (HasText(request.providerKey))
        {
            return new RtcProviderSelection(
                request.providerKey!.Trim(),
                RtcProviderSelectionSource.provider_key
            );
        }

        if (HasText(request.tenantOverrideProviderKey))
        {
            return new RtcProviderSelection(
                request.tenantOverrideProviderKey!.Trim(),
                RtcProviderSelectionSource.tenant_override
            );
        }

        if (HasText(request.deploymentProfileProviderKey))
        {
            return new RtcProviderSelection(
                request.deploymentProfileProviderKey!.Trim(),
                RtcProviderSelectionSource.deployment_profile
            );
        }

        return new RtcProviderSelection(
            defaultProviderKey,
            RtcProviderSelectionSource.default_provider
        );
    }

    private static bool HasText(string? value) => !string.IsNullOrWhiteSpace(value);
}
